documentation/design/rar-core-scenarios.md

```diff
@@ -0,0 +1,193 @@
+outside of the primary set, means that there is no need to cache anything about these assemblies. Everything RAR needs comes (or can come if it's not there already)
+from the `Assemblies` parameter, explicitly provided on each invocation. Note, it may make sense to keep a cache in memory but it definitely doesn't make sense
+to save it to disk.
+
+If we do this, then in the warm and hot scenarios where the per project disk cache exists, we use it only to cache data about NuGet references and project references,
+significantly reducing its size. By eliminating per-reference I/O for most references, RAR would see a significant performance boost.
+
+This is assuming we trust the SDK that it passes correct data and we trust the user that they don't delete or overwrite their SDK files. If this assumption is not
+valid, the mitigation would be to store and check the timestamp of each individual file. We would still benefit from smaller on disk caches, being able to store only
+the timestamp and not assembly name for intact SDK references, but the hot scenario wouldn't get any faster than today.
+
+### Treat NuGet references as immutable
+
+NuGet references live in the NuGet cache which is conceptually immutable. If RAR takes advantage of this, it can eliminate timestamp checks for NuGet references as
+well. The risk is higher than for SDK references because overwriting files in the NuGet cache is commonly used as a cut-the-corner workaround. The benefit is smaller
+because the number of NuGet references is typically lower. The proposal is to use the immutability assumption by default, and provide an opt-out mechanism, likely in
+the form of a new `PackageReference` metadatum for the user to indicate that they plan to update references in the package.
+
+### Don't load the per project disk cache when not needed
+
+As described above, the on disk cache is not adding any value in the hot scenario because its contents already lives in the in-memory cache. The proposal is to
+load it lazily only when (and if) RAR runs into an assembly that does not have a record in the in-memory cache. In developer inner loop, when the same solution is
+built over and over again, the cache would typically not be loaded at all, unless the developer makes a change that actually changes the dependency graph.
+
+As for saving the per-project cache, the logic would stay the same as it is today. If the in-memory cache is marked _dirty_ because it's not identical to what had
+been read from the cache, it would be written back to disk. As a special case, the cache would be considered dirty if the file does not exist at all. This is to
+support scenarios like rebuilding a project by a _hot_ RAR. RAR would be able to satisfy everything from memory, but because rebuild has cleaned the intermediate
+directory, the state would not otherwise be persisted.
```

Comment:
I'm a little confused here. If we no longer read the serialized cache at the start of RAR, then how would we know if it's dirty or not? During RAR's execution, we'd add a bunch of things to the "will be serialized to disk" cache (because it was initially empty), which means it's always dirty.

Reply:
Thank you, these are great points.

My hope is that by implementing other optimizations, primarily the better handling of SDK assemblies, the 70% figure will get significantly lower, making disk cache manipulation more interesting to optimize.

What's still troublesome about the lazy cache saving is that it is non-deterministic. Having a cache that's guaranteed to have certain well-defined content after each build is a very good property to have. For instance, in dev box scenarios it would otherwise be hard to reliably "prime" a repo enlistment - the system may prime by building the full solution and then the developer uses the box to build a specific project that happens to have an incomplete cache and get sub-optimal first-time build perf.

Ok, I think I'm now leaning towards saving always. What we could do perf-wise is perform the save asynchronously. That could have measurable effect on slower storage systems but it may as well be just an unnecessary overcomplication, introducing other issues (e.g. cache read by a process while still being written by another one).

Reply:
I have rewritten the per project disk cache paragraphs. In addition to what's discussed in this comment thread, I've also added a potential optimization to skip saves if we know they won't change the contents of the cache file.